keps/sig-instrumentation/3498-extending-stability/README.md

```diff
@@ -0,0 +1,179 @@
+- `Beta` - representing a more mature stage in a feature metric, with greater stability guarantees than alpha or internal metrics, but less than `Stable`
+
+We also propose amending the semantic meaning of an `Alpha` metric such that it represents the nascent stage of a KEP-proposed feature, rather than the entire class of metrics without stability guarantees. 
+
+Additionally we propose forced upgrades of metrics stability classes in the similar vein that features are not allowed to languish in `alpha` or `beta` stages, but this feature will not be available until the beta version of this KEP. For the alpha version of this KEP, we will implement the necessary changes to Kubernetes metrics framework, such that it supports the additional classes of metrics, without making changes to any existing metrics or their stability levels. As such, this KEP proposes changes to the metrics pipeline and the static analysis pieces of Kubernetes metrics framework. 
+
+
+### Risks and Mitigations
+
+The primary risk is that these changes break our existing (and working) metrics infrastructure. The mitigation should straightfoward, i.e. rollback the changes to the metrics framework. 
+
+## Design Details
+
+Our plan is to add functionality to our static analysis framework which is hosted in the main `k8s/k8s` repo, under `test/instrumentation`. Specifically, we will need to support:
+
+- parsing variables
+- multi-line strings
+- evaluating buckets
+- buckets which are defined via variables and consts
+- evaluation of simple consts
+- evaluation of simple variables
+
+We will not attempt to parse metrics which:
+
+- are constructed dynamically, i.e. through function calls which use function arguments as parameters in metric definitions, since some of those cannot be resolved until runtime. 
+- are constructed using custom prometheus collectors, for the same reasons as above. 
+
+As an aside, much of this work has already been done, but is stashed in a local repo. 
+
+### Test Plan
+
+We have static analysis testing for stable metrics, we will extend our test coverage 
+to include metrics which are `ALPHA` and `BETA` while ignoring `INTERNAL` metrics.
+
+
+### Graduation Criteria
+
+
+#### Alpha
+
+- Kubernetes metrics framework will be enhanced to support additional stability classes
+- The static analysis pipeline of the metrics framework will be enhanced to understand how to parse more things (these are listed above)
+- All instances of `Alpha` metrics will be converted to `Internal`
+
+#### Beta
+
+- Kubernetes metrics framework will be enhanced to support marking `Alpha` and `Beta` metrics with a date. The semantics of this are yet to be determined. This date will be used to statically determine whether or not that metric should be decrepated automatically or promoted.
+- Kubernetes metrics framework will be enhanced with a script to auto-deprecate metrics which have passed their window of existence as an `Alpha` or `Beta` metric
```

Comment:
could we detail the `Alpha` and `Beta` graduation/deprecation timeline as part of this KEP?

Reply:
It's literally a non-goal.